streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java

```diff
@@ -0,0 +1,287 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Arrays;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+public class StreamsKafkaClient {
+
+    private static final Logger log = LoggerFactory.getLogger(StreamsKafkaClient.class);
+    private Metadata metadata;
+    private Metrics metrics;
+    private Selector selector;
+    private ChannelBuilder channelBuilder;
+    private KafkaClient kafkaClient;
+    private StreamsConfig streamsConfig;
+    private Node brokerNode;
+
+    public static final String CLEANUP_POLICY_PROP = "cleanup.policy";
+    private static final String COMPACT = "compact";
+
+    private int maxIterations = 10;
+
+    public StreamsKafkaClient(StreamsConfig streamsConfig) {
+
+        this.streamsConfig = streamsConfig;
+        Time time = new SystemTime();
+
+        Map<String, String> metricTags = new LinkedHashMap<>();
+        metricTags.put("client-id", StreamsConfig.CLIENT_ID_CONFIG);
+
+        this.metadata = new Metadata(streamsConfig.getLong(streamsConfig.RETRY_BACKOFF_MS_CONFIG), streamsConfig.getLong(streamsConfig.METADATA_MAX_AGE_CONFIG));
+        List<InetSocketAddress> addresses = ClientUtils.parseAndValidateAddresses(streamsConfig.getList(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG));
+        metadata.update(Cluster.bootstrap(addresses), 0);
+
+        MetricConfig metricConfig = new MetricConfig().samples(streamsConfig.getInt(CommonClientConfigs.METRICS_NUM_SAMPLES_CONFIG))
+                .timeWindow(streamsConfig.getLong(CommonClientConfigs.METRICS_SAMPLE_WINDOW_MS_CONFIG), TimeUnit.MILLISECONDS)
+                .tags(metricTags);
+        List<MetricsReporter> reporters = streamsConfig.getConfiguredInstances(ProducerConfig.METRIC_REPORTER_CLASSES_CONFIG,
+                MetricsReporter.class);
+        // TODO: This should come from the KafkaStream
+        reporters.add(new JmxReporter("kafka.streams"));
+        this.metrics = new Metrics(metricConfig, reporters, time);
+
+        this.channelBuilder = ClientUtils.createChannelBuilder(streamsConfig.values());
+
+        selector = new Selector(streamsConfig.getLong(streamsConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG), this.metrics, time, "kafka-client", this.channelBuilder);
+
+        kafkaClient = new NetworkClient(
+                selector,
+                metadata,
+                streamsConfig.getString(StreamsConfig.CLIENT_ID_CONFIG),
+                100, // a fixed large enough value will suffice
+                streamsConfig.getLong(streamsConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                streamsConfig.getInt(streamsConfig.SEND_BUFFER_CONFIG),
+                streamsConfig.getInt(streamsConfig.RECEIVE_BUFFER_CONFIG),
+                streamsConfig.getInt(streamsConfig.REQUEST_TIMEOUT_MS_CONFIG), time);
+
+        brokerNode = kafkaClient.leastLoadedNode(new SystemTime().milliseconds());
```

Comment:
I don't think this should be a field. It probably needs to be done before each request as
1. the least loaded node will change.
2. this node might not be up when the request is made.